internal/k8s/owner_fetcher.go

```diff
@@ -44,20 +47,67 @@ func (t ObjectRefTree) String() string {
 	return strings.Join(t.stringLines(), "\n")
 }
 
+type resourceNamespace struct {
+	Namespace Namespace
+	GVK       schema.GroupVersionKind
+}
+
 type OwnerFetcher struct {
-	kCli  Client
-	cache map[types.UID]*objectTreePromise
-	mu    *sync.Mutex
+	globalCtx context.Context
+	kCli      Client
+	cache     map[types.UID]*objectTreePromise
+	mu        *sync.Mutex
+
+	metaCache       map[types.UID]ObjectMeta
+	resourceFetches map[resourceNamespace]*sync.Once
 }
 
-func ProvideOwnerFetcher(kCli Client) OwnerFetcher {
+func ProvideOwnerFetcher(ctx context.Context, kCli Client) OwnerFetcher {
 	return OwnerFetcher{
-		kCli:  kCli,
-		cache: make(map[types.UID]*objectTreePromise),
-		mu:    &sync.Mutex{},
+		globalCtx: ctx,
+		kCli:      kCli,
+		cache:     make(map[types.UID]*objectTreePromise),
+		mu:        &sync.Mutex{},
+
+		metaCache:       make(map[types.UID]ObjectMeta),
+		resourceFetches: make(map[resourceNamespace]*sync.Once),
 	}
 }
 
+func (v OwnerFetcher) getOrCreateResourceFetch(gvk schema.GroupVersionKind, ns Namespace) *sync.Once {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	rns := resourceNamespace{Namespace: ns, GVK: gvk}
+	fetch, ok := v.resourceFetches[rns]
+	if !ok {
+		fetch = &sync.Once{}
+		v.resourceFetches[rns] = fetch
+	}
+	return fetch
+}
+
+// As an optimization, we batch fetch all the ObjectMetas of a resource type
+// on startup.
```

Comment:
Maybe nitty, but this isn't at startup, is it? IIUC, it's the first time we check the objectmeta of any instance of a type, which I think is, like, the first time event watcher or pod watcher get updates? In either case, this part of the code is sufficiently removed from whether it happens at startup that the comment is hard to verify / keep up to date.

Reply:
good point! I reworded this